### ISSUE TITLE
"rule-non-nested-empty-line-before" not ignoring first node?

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#507]
Ref: https://github.com/stylelint/stylelint/issues/504#issuecomment-157339912

> With `"rule-non-nested-empty-line-before": [2, "always"]`

> I always get an error for the first line of my css (because there is no
space at the beginning of the rule at the start of the file). Is there a
way to avoid that?

@raynor85 This shouldn't be happening as the first node in a file is ignored by the rule. Can you please paste the top few lines of your CSS file here so we can work out what's going on?

**Comment by raynor85** [comment]
This affected all the css, like this one:

.bauer-header {
    height: 200px;
}

In the meanwhile I was able to solve both problems. Both issues appeared
with this config file:

{
    "rules": {
        "color-no-invalid-hex": 2,
        "block-opening-brace-newline-after": [2, "always"],
        "declaration-colon-space-after": [2, "always"],
        "declaration-colon-space-before": [2, "never"],
        "function-comma-space-after": [2, "always"],
        "indentation": [2, 4],
        "rule-no-single-line": 2,
        "rule-non-nested-empty-line-before": [2, "always"],
        "rule-trailing-semicolon": [2, "always"],
        "selector-combinator-space-after": [2, "always"],
        "selector-list-comma-newline-after": [2, "always"],
        "selector-no-id": 2
    }
}

Changing it as the following solved both problems:

{
    "rules": {
        "color-no-invalid-hex": 2,
        "block-opening-brace-newline-after": [2, "always"],
        "declaration-colon-space-after": [2, "always"],
        "declaration-colon-space-before": [2, "never"],
        "function-comma-space-after": [2, "always"],
        "indentation": [2, 4],
        "rule-nested-empty-line-before": [2, "always", {"except":
["first-nested"], "ignore": ["after-comment"]}],
        "rule-no-single-line": 2,
        "rule-non-nested-empty-line-before": [2, "always", {"ignore":
["after-comment"]}],
        "rule-trailing-semicolon": [2, "always"],
        "selector-combinator-space-after": [2, "always"],
        "selector-list-comma-newline-after": [2, "always"],
        "selector-no-id": 2
    }
}

**Comment by jeddy3** [comment]
I'm glad you got it working, but that change to the config doesn't explain why it wasn't working as expected without the `ignore` optional option. And I wouldn't mind getting to the bottom of it, if you don't mind helping me out.

All I can think of is if the first rule in the file was preceded by a comment. Is this _exactly_ what you have at the beginning of your file?:

```css
.bauer-header {
    height: 200px;
}
```

Or did a comment come before it? e.g.

```
/* some comment */
.bauer-header {
    height: 200px;
}
```

Thanks :)

**Comment by raynor85** [comment]
I am very glad to help you out, that is why I posted my solution (maybe it
can help as well).

No comment before, just the plain text.

Ruben

**Comment by jeddy3** [comment]
> No comment before, just the plain text.

Oh well. There's a passing test for it in https://github.com/stylelint/stylelint/blob/master/src/rules/rule-non-nested-empty-line-before/__tests__/index.js#L12

So unless @davidtheclark wishes to investigate further, I'm going to close this and keep it in mind should something similar to this crop up again.

**Comment by esseb** [comment]
I could reproduce this issue using `webpack`, `postcss`, and `postcss-import` using a `main.css` that looks like this:
```
@charset "UTF-8";

@import "utils/*.css";
@import "base/*.css";
@import "modules/*.css";
```

Stylelint insisted the first line of `base/base.css` was invalid:
```
html {
  box-sizing: border-box;
}
```

The issue disappeared after I changed `main.css` to this:
```
@charset "UTF-8";

@import "utils/*.css";

@import "base/*.css";
@import "modules/*.css";
```

Not entirely sure what in the `utils` folder caused the issue, and modifying those files did not make the problem disappear for me.

I've worked around the issue for now in my project by moving the stylelint plugin about postcss-importer.

Leaving this comment in here in case anyone else encounters this issue.

**Comment by davidtheclark** [comment]
I think that the issue is that you are running stylelint *after* postcss-import; and postcss-import is concatenating your files together without an empty-line between files; so the CSS that stylelint is actually looking at *does* violate the rule.

We recommend running stylelint either before other plugins or as part of a separate process that does not involve transforms.

**Comment by jeddy3** [comment]
> Leaving this comment in here in case anyone else encounters this issue.

Thanks for that :)

> I've worked around the issue for now in my project by moving the stylelint plugin above postcss-importer.

Putting stylelint before a plugin like `postcss-import` is the recommended way to go about things right now, as a plugin like `postcss-import` might transform your CSS into something that no longer matches your config. 

FYI. In this instance, it looks like `postcss-import` is, correctly, transforming your CSS by inserting the contents of each of your imports into the same sheet, and therefore there is—in this transformed CSS—no empty line between the first rule of the second import and the last rule of the first import.

If you're interested, there's a little more information about this in the [CSS Processors](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/css-processors.md) section of the docs.

**Comment by jeddy3** [comment]
Oops, double post :)

@esseb You might also be interested in following https://github.com/postcss/postcss-import/issues/117 as, I think, the intention there is to allow `stylelint` to be run on each of the imported files _before_ they are concatenated.

But, as we both said above, putting `stylelint` before plugins like `postcss-import` is the way to go right now :)

**Comment by esseb** [comment]
@davidtheclark ah, you're right. I didn't think it was that simple since I didn't get a similar warning for my `modules/button.css` file, but it looks like the `$variables` at the top of that file prevented this rule from applying. I only have about 4 css files in my project so far.

@jeddy3 Hm, `stylelint` will only attempt to lint `main.css` in that case, using my current setup.

**Comment by esseb** [comment]
I'll follow https://github.com/postcss/postcss-import/issues/117, that seems like the proper solution to this.

I could use a similar solution as described in https://github.com/stylelint/stylelint/issues/133#issuecomment-120549292, but I'm not in a rush.

Thanks for the help, @davidtheclark and @jeddy3

**Comment by jeddy3** [comment]
No problem :)
